Ucsf landing page UI

schroerbrian opened the pull request:
This PR adds the UI and functionality for the UCSF landing page and the Client Eligibility page (includes desktop and mobile). It uses mock data to populate the resources and eligibilities. This data and rendering logic will be updated when the API endpoints are wired up. I have added a few todos in the code to indicate where the updates will occur.

![image](https://user-images.githubusercontent.com/3012520/171265803-8798b36d-5530-4c15-9d96-5337645431b8.png)

![image](https://user-images.githubusercontent.com/3012520/171265839-383f4c56-4147-41a1-979b-3b6902508955.png)

richardxia commented on app/pages/UcsfHomePage/UcsfHomePage.module.scss:
```diff
@@ -0,0 +1,62 @@
+  }
+}
+
+.resourceLabel {
```
The Figma design shows that the text in the label should have a font size of 18px, but because you don't set it here, it ends up being 16px, which is a tiny bit smaller that it's supposed to be.

I'm also seeing that the Figma design chooses a font weight of 500, but we don't have that font size available; the three font weights that we currently download are 400, 600, and 700. We may want to clarify with the design team exactly which font faces we're supposed to be using.

richardxia commented on app/pages/UcsfHomePage/UcsfHomePage.tsx:
```diff
@@ -0,0 +1,102 @@
+  resourceList: resourceListItem[];
+  setResourceList: (resourcesArray: resourceListItem[]) => void;
+}) => {
+  const updateCheckedResources = (evt: React.ChangeEvent<HTMLInputElement>) => {
+    const newList = resourceList.map(item => {
+      if (item.name === evt.target.name) {
+        const updatedItem = {
+          ...item,
+          checked: !item.checked,
+        };
+
+        return updatedItem;
+      }
+
+      return item;
+    });
+
+    setResourceList(newList);
+  };
```
I think it might be a bit cleaner to decompose this into two separate functions:

1. One that allows you to transform a single item in the array based on its array index
2. One that actually toggles the `checked` property on the resource list item

Once you do that, then when you actually render the HTML a bit lower, in the `resourceList.map()` call, use the two-argument form of the mapped function to get the index of the element being mapped, and then use that with function 1) described above.

Here's a sketch of what I'm suggesting:

```js
const setResourceItem = (index, newResource) => {
  const newList = [
    ...resourceList.slice(0, index),
    newResource,
    ...resourceList.slice(index + 1, -1),
  ];
  setResourceList(newList);
};

const toggleChecked = (resource) => { /* ... */ };

return (
  <ul>
    {resourceList.map((resource, i) => (
      <button onChange={() => setResourceItem(i, toggleChecked(resource))} />
    )}
  </ul>
);
```

schroerbrian replied:
Cleaner and more efficient. Thanks!

richardxia reviewed (commented):
Thanks again for the work on this! I haven't gotten a chance to look over the `UcsfClientEligibilityPage`, but I hope the comments that I've left so far are helpful. Let me know if you have any questions about them.

schroerbrian commented on app/App.tsx:
```diff
@@ -53,6 +57,14 @@ const outerContainerId = 'outer-container';
+
+  let homePageComponent = HomePage;
+  if (checkWhiteLabelPage('ucsf')) {
+    homePageComponent = UcsfHomePage;
```
That is a very sound suggestion. I have removed the ID and added a "homePageComponent" property to the whitelabel object.

schroerbrian commented on app/components/ui/inline/Input/Input.tsx:
```diff
@@ -0,0 +1,38 @@
+    checked={checked}
+    className={`${styles.inputDefaults} ${addClass || ''}`}
+    id={id}
+    type={type}
```
Yep, that makes sense. I've gone ahead and changed this to Checkbox.tsx

schroerbrian commented on app/components/ui/inline/Input/Input.module.scss:
```diff
@@ -0,0 +1,11 @@
+  &[type=checkbox] {
+    cursor: pointer;
+    transform: scale(1.66);
+  }
```
I've gone ahead and changed this, largely using the example of a resource that you shared: https://moderncss.dev/pure-css-custom-checkbox-style/.  (I'd tried this previously but got in a rabbit hole and moved on and did it the easy way). This does take a good chunk of CSS to get it done, but looks and works pretty nicely.

schroerbrian commented on app/styles/utils/_mixins.scss:
```diff
@@ -282,4 +282,21 @@ $prefix-for-spec:      true !default; // required for keyframe mixin
+@mixin buttonLarge($color: $color-white, $bgColor: $color-brand) {
+  @include buttonBasic($color, $bgColor);
+  width: 225px;
+}
```
Yep, I have create a component called TileButton (kind of awkward but maybe a little more descriptive than just Tile or Button?)

schroerbrian commented on app/styles/components/_ucsf-layout.scss:
```diff
@@ -0,0 +1,17 @@
+  width: 90%;
+}
+
+.title {
+  @extend %font-size-large;
+  color: $color-title;
+}
+
+.subTitle {
+  @extend %font-size-medium;
+  color: $color-grey6;
+  margin: 12px 0 22px;
+}
```
> Thanks again for the work on this! I haven't gotten a chance to look over the `UcsfClientEligibilityPage`, but I hope the comments that I've left so far are helpful. Let me know if you have any questions about them.

Thanks! They are very helpful. Thanks for your thoroughness. I have wondered just how granular one should make React components and I'm beginning to see that they should be pretty granular, which I think is great. How about you hold off on reviewing the `UcsfClientEligibilityPage` so that I can apply some of the lessons learned from these comments to that component before you look at it so as to save you some time. I have not been following the the Figma designs with the greatest fidelity, so I will fix that up for both pages and make other changes.

schroerbrian commented:
> Thanks again for the work on this! I haven't gotten a chance to look over the `UcsfClientEligibilityPage`, but I hope the comments that I've left so far are helpful. Let me know if you have any questions about them.

Thanks! They are very helpful. Thanks for your thoroughness. I have wondered just how granular one should make React components and I'm beginning to see that they should be pretty granular, which I think is great. How about you hold off on reviewing the UcsfClientEligibilityPage so that I can apply some of the lessons learned before you look at it so as to save you some time. I have not been following the the Figma designs with the greatest fidelity, so I will fix that up for both pages and make other changes.

I am still working on your last few suggestions for the `UcsfHomePage`. I''ll probably finish those up by Wednesday eve.

schroerbrian commented on app/styles/components/_ucsf-layout.scss:
```diff
@@ -0,0 +1,17 @@
+.ucsfHomePage {
+  margin: 100px auto 0;
+  width: 90%;
+}
```
This is a great way for pixel perfect UI. I moved the .ucsfHomePage style into a layout component that acts as a simple grid wrapper for the UcsfHomePage and UcsfClientEligibilityPage

schroerbrian commented on app/pages/UcsfHomePage/UcsfHomePage.module.scss:
```diff
@@ -0,0 +1,62 @@
+  height: 160px;
+  margin: 0 15px 15px 0;
+  border-radius: 5px;
+  box-shadow: 0 0 4px rgb(0 0 0 / 20%);
```
Changed. I was erring towards copying existing CSS styles that were already in our code base (we had a box shadow value being used already that I copied) vs fidelity to the UCSF designs, but it seems like we may want to go for a clean break.

schroerbrian commented on app/pages/UcsfHomePage/UcsfHomePage.tsx:
```diff
@@ -0,0 +1,102 @@
+    <ul className={styles.resourceList}>
+      {resourceList.map((resource: resourceListItem) => (
+        <li
+          key={resource.name}
```
Done and agreed. I think I had just used the resource.name prior to adding the id property to my mock data.

schroerbrian commented on app/pages/UcsfHomePage/UcsfHomePage.module.scss:
```diff
@@ -0,0 +1,62 @@
+.resourceLabel {
+  height: 100%;
+  width: 100%;
+  margin-top: -10px;
```
Fixed and simplified. Good eye! The checkbox component now has styles that are consistent across browsers. It was inspired by: https://moderncss.dev/pure-css-custom-checkbox-style/

schroerbrian commented:
> > Thanks again for the work on this! I haven't gotten a chance to look over the `UcsfClientEligibilityPage`, but I hope the comments that I've left so far are helpful. Let me know if you have any questions about them.
> 
> Thanks! They are very helpful. Thanks for your thoroughness. I have wondered just how granular one should make React components and I'm beginning to see that they should be pretty granular, which I think is great. How about you hold off on reviewing the UcsfClientEligibilityPage so that I can apply some of the lessons learned before you look at it so as to save you some time. I have not been following the the Figma designs with the greatest fidelity, so I will fix that up for both pages and make other changes.
> 
> I am still working on your last few suggestions for the `UcsfHomePage`. I''ll probably finish those up by Wednesday eve.

I've fixed up all of your suggestions for UCSF. Thanks again, these are very good, helpful suggestions! 

I hadn't been sure whether to err on using styles that we already use on the site for checkboxes, inputs, etc., vs. creating new styles/components based on these designs. But after reviewing your comments and updating the code, it seems like creating new styles and components is better than using/mimicking some of our existing legacy UI. 

I have updated some of the UCSF Client Eligibility page code based on your Home Page suggestions, but I am not yet done with that. I will let you know when I think that is good to review as well.

schroerbrian commented:
OK, I think both the UCSF Home Page AND Client Eligibility Page are in a good spot to be reviewed, at last. Let me know how it's looking and thanks for your time, @richardxia !

richardxia commented on app/pages/UcsfClientEligibilityPage/UcsfClientEligibilityPage.module.scss:
```diff
@@ -0,0 +1,108 @@
+
+.eligibilityLabel {
+  cursor: pointer;
+  vertical-align: top;
```
Hmm, I don't have a quick fix for this, but if you set the Checkbox's vertical-align to `middle` instead of `bottom` like it currently is, then you can also set the `vertical-align` for this element to `middle`, which will actually middle-align both of them.

Vertically aligning inline content is pretty tricky in general due to Western languages having descenders (e.g. the tail of a lowercase "g"). You probably don't need to read this right now, but I thought [this article](https://christopheraue.net/design/vertical-align) was one of the best at explaining how vertical alignment actually works for inline content, and the short summary is that it's really, really complicated and depends on the vertical alignment of all other items on the same line.

Once I spent a long time fiddling with a custom checkbox and vertical alignment of a label next to it, and I think what I ended up doing was keeping the default vertical alignment of `baseline` for both the checkbox and the label, but instead carefully positioning the checkbox vertically such that the baseline of checkbox was not at the bottom edge but instead slightly higher up. I also set the checkbox height and width in `em` so that it scales with the font size, which is important in terms of lining up the descender height of a typical Latin font with the checkbox's lower edge. The other really tricky thing is that this is also dependent on your font, since different fonts have different descender, x, and topline heights.

Anyway, I don't think we need to really change this now, but I wanted to point out a potential refactoring we may want to do in the future.

richardxia reviewed (commented):
Nice, this is looking pretty good to me! I think most of my comments are relatively minor, and probably the larger comments are related to the layout of the second page, with the eligibilities box.

I'll ask the question about the layout of the next/back buttons on Figma that I said I would in one of my comments.

Aside from that, one other thing I noticed while playing with this locally is that the first page doesn't require you to select any options. Based on the Figma design, it looks like the "Next" button is disabled and slightly faded out when nothing is selected. This doesn't necessarily need to be fixed here, since it looks like we'll need to actually plug in the logic in a followup PR, but we'll probably need to add some more styling in addition to the logic in order to make it look disabled.

schroerbrian commented on app/components/ucsf/Section/Section.tsx:
```diff
@@ -0,0 +1,20 @@
+  body?: string;
+  addClass?: string;
+}) => (
+  <section className={`${styles.section} ${addClass || ''}`}>
```
Good catch! I switched on the `restrict-template-expressions` and saw that it complains about `style.className` expressions that map to valid styles. It gives the following error: `Invalid type "any" of template literal expression`. We have quite a few of these, so I left this off for now. I guess that is not typesafe as it could be an undefined element. Any thoughts on this?

schroerbrian commented on app/pages/UcsfHomePage/UcsfHomePage.module.scss:
```diff
@@ -0,0 +1,64 @@
+  border-radius: 8px;
+  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1), 0px 0px 4px rgba(0, 0, 0, 0.1);
+  &.isChecked {
+    border: solid 1px $color-brand;
```
Setting the border to transparent for unchecked works, thanks!

schroerbrian commented on app/pages/UcsfClientEligibilityPage/UcsfClientEligibilityPage.module.scss:
```diff
@@ -0,0 +1,108 @@
+
+.eligibilitiesContainer {
+  display: grid;
+  grid-template-columns: repeat(22, 1fr);
```
I used grid 22 for a placing an element in the middle of a column.  (Since this page was taking up an 11 span, I had to use 22 to get the math right). It was a little overcomplicated to manage this though.

I guess the point is moot given we now are using an updated design, and I have removed the 22 column grid!

schroerbrian commented on app/components/ui/inline/TileButton/TileButton.tsx:
```diff
@@ -0,0 +1,38 @@
+     // limited to ButtonType enums, commenting this out should be safe
+     /* eslint-disable-next-line react/button-has-type */
+    type={buttonType}
+    className={`${size ? styles[size] : ''} ${addClass || ''}`}
```
Size is an optional param, so TS was complaining when I removed the `size` check. But fortunately this is also moot due to your good suggestion below to remove the size variants

schroerbrian commented on app/components/ui/inline/TileButton/TileButton.module.scss:
```diff
@@ -0,0 +1,19 @@
+.large {
+  @include buttonBasic;
+  width: 218px;
+}
```
Makes great sense. I set the button min-width to be 120px (the size of the "Next" and "Back" buttons in the designs), and added enough padding to make the large button to be approximately 218px width (also matching the designs). Great suggestion, thanks!

schroerbrian commented on app/pages/UcsfClientEligibilityPage/UcsfClientEligibilityPage.tsx:
```diff
@@ -0,0 +1,170 @@
+                <li key={`${eligibilityItem.label}-${eligibility.name}`} className={styles.eligibilityItem}>
+                  <Checkbox
+                    onChange={() => setEligibilityItem(index, toggleChecked(eligibilityItem, i))}
+                    name={eligibility.name}
```
Thank you, I was struggling with the var naming here. `EligibilityGroup` is much better. Fixed!

schroerbrian commented on app/pages/UcsfClientEligibilityPage/UcsfClientEligibilityPage.module.scss:
```diff
@@ -0,0 +1,108 @@
+
+.subtitleMargin {
+  margin-bottom: 13px;
+}
```
Unnecessary, I must have missed the .elgibilitiesBox margin being already set.

schroerbrian commented on app/pages/UcsfClientEligibilityPage/UcsfClientEligibilityPage.module.scss:
```diff
@@ -0,0 +1,108 @@
+.eligibilitiesLabels {
+  display: grid;
+  gap: 45px;
+  list-style: decimal;
```
DIdn't know once could set multiple `list-styles`!

schroerbrian commented:
> Nice, this is looking pretty good to me! I think most of my comments are relatively minor, and probably the larger comments are related to the layout of the second page, with the eligibilities box.
> 
> I'll ask the question about the layout of the next/back buttons on Figma that I said I would in one of my comments.
> 
> Aside from that, one other thing I noticed while playing with this locally is that the first page doesn't require you to select any options. Based on the Figma design, it looks like the "Next" button is disabled and slightly faded out when nothing is selected. This doesn't necessarily need to be fixed here, since it looks like we'll need to actually plug in the logic in a followup PR, but we'll probably need to add some more styling in addition to the logic in order to make it look disabled.

Thanks again! I have a mental note to disable the button on the first page if no options are selected. I also think that we may want to auto-check all the options of an eligibility group on the 2nd page when "See All" is selected. I will ask the designers about that.

richardxia reviewed (approved):
LGTM! Thanks for getting this through, it's looking pretty slick to see this come to life when running the branch locally!

schroerbrian commented:
Thanks again for all of your helpful input!
